File: go/src/offertwentyfourth/offertwentyfourth.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil{
		return head
	}
	rhead := &ListNode{}
	head = head.Next
	cur := head
	for cur != nil {
		tmp := cur.Next
		cur.Next = rhead.Next
		rhead.Next = cur
		cur = tmp
	}
	return rhead
}

func printList(head *ListNode) {
	head = head.Next
	for head != nil {
		fmt.Println(head.Val)
		head = head.Next
	}
}

func constructList(arr []int) *ListNode {
	head := &ListNode{}
	p := head
	for i := 0; i < len(arr); i++ {
		node := &ListNode{Val:arr[i]}
		p.Next = node
		p = p.Next
	}
	return head
}

func main() {
	arr := []int{1, 2, 3, 4, 5, 6, 7}
	ln := constructList(arr)
	printList(ln)
	res := reverseList(ln)
	printList(res)
}

File: c++/offertwentyfourth/offertwentyfourth.cpp
#include <iostream>
#include <vector>
using namespace std;

struct ListNode{
    int val;
    ListNode *next;
    ListNode(int x){val=x;}
};

ListNode* reverseList(ListNode *head){
    // ListNode *rhead = new ListNode(0);
    // rhead->next = NULL;
    // while (head != NULL) {
    //     ListNode *node = new ListNode(head->val);
    //     node->next = rhead->next;
    //     rhead->next = node;
    //     head = head->next;
    // }
    // return rhead;
    ListNode *pre = NULL, *cur = head;
    while (cur) {
        ListNode* temp = cur->next;
        cur->next = pre;
        pre = cur;
        cur = temp;
    }
    return pre;
}

void printList(ListNode *head) {
    while (head != NULL) {
        cout<<head->val<<" ";
        head = head->next;
    }
}

ListNode* constructList(vector<int> arr) {
    ListNode *head = NULL;
    ListNode *tail = head;
    for (int i=0; i<arr.size(); i++) {
        ListNode *node = new ListNode(arr[i]);
        tail->next = node;
        tail = tail->next;
    }
    return head;
}

int main(){
    vector<int> arr = {1, 2, 3, 4, 5, 6, 7};
    ListNode *head = constructList(arr);
    printList(head);
    ListNode *res = reverseList(head);
    printList(res);
    system("pause");
    return 0;
}